src/PowerShellEditorServices/Language/TokenOperations.cs

```diff
@@ -21,100 +20,65 @@ internal static class TokenOperations
-            // that the largest range (i.e. most number of lines spanned) is sorted
-            // first. This is needed to detect duplicate regions. The first in the list
-            // will be used and subsequent duplicates ignored.
-            foldableRegions.Sort();
-
-            // It's possible to have duplicate or overlapping ranges, that is, regions which have the same starting
-            // line number as the previous region. Therefore only emit ranges which have a different starting line
-            // than the previous range.
-            foldableRegions.RemoveAll( (FoldingReference item) => {
-                // Note - I'm not happy with searching here, but as the RemoveAll
-                // doesn't expose the index in the List, we need to calculate it. Fortunately the
-                // list is sorted at this point, so we can use BinarySearch.
-                int index = foldableRegions.BinarySearch(item);
-                if (index == 0) { return false; }
-                return (item.StartLine == foldableRegions[index - 1].StartLine);
-            });
+            for (int index = 0; index < tokens.Length; index++)
+            {
+                Token token = tokens[index];
+                if (token.Kind == TokenKind.Comment) {
+                    // Processing for comment regions <# -> #>
+                    if ((token.Kind == TokenKind.Comment) && (token.Extent.StartLineNumber != token.Extent.EndLineNumber)) {
+                        foldableRegions.Add(CreateFoldingReference(token, token, RegionKindComment));
+                    }
 
-            // Some editors have different folding UI, sometimes the lastline should be displayed
-            // If we do want to show the last line, just change the region to be one line less
-            if (ShowLastLine) {
-                foldableRegions.ForEach( item => { item.EndLine--; });
+                    if (IsBlockComment(index, tokens)) {
+                        // Regex's are very expensive.  Use them sparingly.
+                        bool isStartRegion = s_startRegionTextRegex.IsMatch(token.Text);
+                        bool isEndRegion = s_endRegionTextRegex.IsMatch(token.Text);
+                        // Processing for # region -> # endregion
+                        if (isStartRegion) {
+                            tokenCommentRegionStack.Push(token);
+                        }
+                        if ((tokenCommentRegionStack.Count > 0) && isEndRegion) {
+                            foldableRegions.Add(CreateFoldingReference(tokenCommentRegionStack.Pop(), token, RegionKindRegion));
+                        }
+                        // If it's neither a start or end regsion then it could be block line comment
+                        // Processing for blocks of line comments # comment1\n# comment2\n...
+                        if (!(isStartRegion || isEndRegion)) {
```

Comment:
`continue`s above will mean not needing an `if` here

Reply:
Fixed with some other modifications to the if statement above.